src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/SelfHosted/SelfHostedSyncOrganizationSponsorshipsCommand.cs

```diff
@@ -0,0 +1,108 @@
+
+        public SelfHostedSyncOrganizationSponsorshipsCommand(
+        IOrganizationSponsorshipRepository organizationSponsorshipRepository,
+        IOrganizationUserRepository organizationUserRepository,
+        ILicensingService licensingService,
+        IGlobalSettings globalSettings,
+        ILogger<SelfHostedSyncOrganizationSponsorshipsCommand> logger) : base("vault.bitwarden.com", "identity.bitwarden.com", "api.installation", globalSettings.Installation.Id.ToString(), globalSettings.Installation.Key, logger)
+        {
+            _organizationUserRepository = organizationUserRepository;
+            _organizationSponsorshipRepository = organizationSponsorshipRepository;
+            _licensingService = licensingService;
+        }
+
+        public async Task SyncOrganization(Guid organizationId)
+        {
+            var billingSyncKey = await GetBillingSyncKey(organizationId);
+
+            if (string.IsNullOrWhiteSpace(billingSyncKey))
+            {
+                throw new BadRequestException($"No Billing Sync Key known for organization {organizationId}");
+            }
+
+            var cloudOrganizationId = (await _licensingService.ReadOrganizationLicenseAsync(organizationId)).Id;
+            var orgUsers = await _organizationUserRepository.GetManyByOrganizationAsync(organizationId, null);
+            var organizationSponsorships = await _organizationSponsorshipRepository.GetBySponsoringOrganizationAsync(organizationId);
+
+            foreach (var orgSponsorshipsBatch in CoreHelpers.Batch(organizationSponsorships, 1000))
+            {
+                var response = await SendAsync<OrganizationSponsorshipSyncModel, OrganizationSponsorshipSyncModel>(HttpMethod.Post, "organizationSponsorships/sync", new OrganizationSponsorshipSyncModel
+                {
+                    BillingSyncKey = billingSyncKey,
+                    SponsoringOrganizationCloudId = cloudOrganizationId,
+                    SponsorshipsBatch = orgSponsorshipsBatch.Select(s => new OrganizationSponsorshipModel
+                    {
+                        SponsoringOrganizationUserId = s.SponsoringOrganizationUserId,
+                        FriendlyName = s.FriendlyName,
+                        OfferedToEmail = s.OfferedToEmail,
+                        PlanSponsorshipType = s.PlanSponsorshipType,
+                        // ValidUntil = s.ValidUntil,
+                        // ToDelete = s.ToDelete
+                    })
+                });
+
+                if (response == null) 
+                {
+                    throw new BadRequestException("Organization sync failed");
+                }
+
+                foreach (var sponsorshipModel in response.SponsorshipsBatch)
+                {
+                    var existingOrgSponsorship = await _organizationSponsorshipRepository
+                        .GetBySponsoringOrganizationUserIdAsync(sponsorshipModel.SponsoringOrganizationUserId.GetValueOrDefault());
```

Comment:
We already have a list of sponsorships we're working on, no need to grab it again.

```suggestion
                    var existingOrgSponsorship = orgSponsorshipsBatch.FirstOrDefault(s => s.SponsoringOrganizationUserId == sponsorshipModel.SponsoringOrganizationUserId);
```